### PR TITLE
librbd: new exclusive lock release journal event

### PR DESCRIPTION
**Pull request by sahithi-rv**
Signed-off-by: Sahithi R V <sahithi.rv1@gmail.com>

### CODE REVIEW

**Comment by sahithi-rv** [comment]
The event created is not yet fully functional.

**Comment by dillaman** [comment]
Can you put the tracker # in the commit message:

> Fixes: #14414

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 1]
```diff
@@ -131,9 +132,31 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
 
   assert(*ret_val == 0);
   send_close_journal();
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Remove this line -- we need to append the event before closing the journal

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 2]
```diff
@@ -131,9 +132,31 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+
+  send_append_journal_event();
   return nullptr;
 }
 
+void send_append_journal_event(){
+    CephContext *cct = m_image_ctx.cct;
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Move lines 162 - 171 from ```send_close_journal``` to here.

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 3]
```diff
@@ -131,9 +132,31 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+      klass, &klass::handle_add_journal_event>(this);
+    //create journal event
+    //journal::EventEntry event_entry(journal::ExclusiveLockReleaseEvent());
+    //m_journal->append_op_event(journal_tid,event,)
+}
+
+Context *handle_append_journal_event(int *ret_val){
+    CephContext *cct = m_image_ctx.cct;
+    ldout(cct, 10) << __func__ << ": r=" << *ret_val << dendl;
+
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: If an error, log it to lderr -- not much to recover from at this point.

**Comment by dillaman** [comment]
Also, please add a test case for this new event.  See https://github.com/ceph/ceph/blob/master/src/test/librbd/journal/test_mock_Replay.cc#L454 for an example.

**Comment by jdurgin** [comment]
Looks like this needs rebasing after some recent merges. Could you put the "Fixes: #XYZ" on a separate line in the commit messages, like in: https://github.com/ceph/ceph/commit/5d4173543fe9ee2be0139f228cdc82d3e0135aad ? It helps keep the first line shorter as well

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 4]
```diff
@@ -129,23 +130,50 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
   CephContext *cct = m_image_ctx.cct;
   ldout(cct, 10) << __func__ << dendl;
 
-  assert(*ret_val == 0);
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Still a valuable assert statement

File: src/librbd/exclusive_lock/ReleaseRequest.h [hunk 5]
```diff
@@ -41,7 +41,10 @@ class ReleaseRequest {
    *    v
    * FLUSH_NOTIFIES . . . . . . . . . . . . . .
    *    |                                     .
-   *    v                                     .
+   *    v
+   * ADD_JOURNAL_EVENT
+   *    |
+   *    v                                    .
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.h]: Minor: fix the ASCII art

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 6]
```diff
@@ -129,23 +130,50 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
   return nullptr;
 }
 
-template <typename I>
-void ReleaseRequest<I>::send_close_journal() {
-  {
-    RWLock::WLocker snap_locker(m_image_ctx.snap_lock);
-    std::swap(m_journal, m_image_ctx.journal);
-  }
+void send_append_journal_event(){
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Your new methods should be class members

File: src/librbd/journal/Types.h [hunk 7]
```diff
@@ -253,6 +254,18 @@ struct FlattenEvent : public OpEventBase {
   using OpEventBase::dump;
 };
 
+struct ExclusiveLockReleaseEvent {
+  static const EventType TYPE = EVENT_TYPE_EXCLUSIVE_LOCK_RELEASE;
+
+  ExclusiveLockReleaseEvent() {
+  }
+  ExclusiveLockReleaseEvent(uint64_t op_tid) : OpEventBase(op_tid) {
```
Comment (dillaman) [inline src/librbd/journal/Types.h]: Don't need this constructor -- this doesn't derive from OpEventBase

**Comment by sahithi-rv** [comment]
@dillaman Please test it now.

**Comment by sahithi-rv** [comment]
@dillaman can you please test it now?

**Comment by dillaman** [comment]
@sahithi-rv You can just check the github build status at the bottom of the PR.  It says you need to rebase your PR to the current master branch.

**Comment by sahithi-rv** [comment]
rebased.

**Comment by dillaman** [comment]
Getting closer -- the unit tests are crashing.  You can repeat this locally by running "RBD_FEATURES=109 ./unittest_librbd" from the 'src' directory.

File: src/librbd/journal/Types.h [hunk 8]
```diff
@@ -253,6 +254,18 @@ struct FlattenEvent : public OpEventBase {
   using OpEventBase::dump;
 };
 
+struct ExclusiveLockReleaseEvent : public OpEventBase {
```
Comment (dillaman) [inline src/librbd/journal/Types.h]: Don't need to inherit from OpEventBase -- this isn't a maintenance operation.  As a result, you won't need an op_tid.

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 9]
```diff
@@ -130,22 +131,53 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+    }
+
+    using klass = ReleaseRequest<I>;
+    Context *ctx = create_context_callback<
+      klass, &klass::handle_append_journal_event>(this);
+    //append journal event
+    uint64_t journal_op_tid = m_image_ctx.journal->allocate_op_tid();
+    {
+      RWLock::WLocker snap_locker(m_image_ctx.snap_lock);
+      m_journal->append_op_event(journal_op_tid,journal::EventEntry(journal::ExclusiveLockReleaseEvent(journal_op_tid)),ctx);
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Recommend creating a new method: ```librbd::Journal::append_lock_released_event()``` which injects the event into the journal.

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 10]
```diff
@@ -130,22 +131,53 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+    if (m_journal == nullptr) {
+      send_close_object_map();
+      return;
+    }
+
+    using klass = ReleaseRequest<I>;
+    Context *ctx = create_context_callback<
+      klass, &klass::handle_append_journal_event>(this);
+    //append journal event
+    uint64_t journal_op_tid = m_image_ctx.journal->allocate_op_tid();
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: This is where the unit tests are currently crashing since ```m_image_ctx.journal``` is a nullptr due to the swap on line 145.  Regardless, this will no longer be needed once the new Journal method is added (see below)

**Comment by sahithi-rv** [comment]
@dillaman I don't get any error when I run unittest_librbd on local machine.

File: src/librbd/journal/Types.h (hunk 8 above)
Comment (sahithi-rv) [inline src/librbd/journal/Types.h]: Then what do I do about encode, decode, dump member functions.

Comment (dillaman) [inline src/librbd/journal/Types.h]: Implement them as empty methods just like UnknownEvent

File: src/librbd/exclusive_lock/ReleaseRequest.cc (hunk 10 above)
Comment (sahithi-rv) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: There is no failure when I run it on local machine!

Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Fails in the bot and fails on my machine.  Are you 1000% confident that you have the environment variable "RBD_FEATURES=109" set before running unittest_librbd?   You can clearly see from just the code above that you will be dereferencing a nullptr.

**Comment by sahithi-rv** [comment]
when I run RBD_FEATURES=109 ./unittest_librbd  - The process gets stuck.
err

**Comment by sahithi-rv** [comment]
Can you check the console output. It shows all tests as passed yet throws error.

**Comment by sahithi-rv** [comment]
Can I know what the present error is due to? It says "Connection was broken"

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 11]
```diff
@@ -130,22 +131,52 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+    if (m_journal == nullptr) {
+       send_close_object_map();
+       return;
+    }
+
+    using klass = ReleaseRequest<I>;
+    Context *ctx = create_context_callback<
+      klass, &klass::handle_append_journal_event>(this);
+    //append journal event
+    uint64_t journal_op_tid = m_journal->allocate_op_tid();
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: No need to allocate op tid -- the release lock event isn't an operation.

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 12]
```diff
@@ -130,22 +131,52 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+    {
+      RWLock::WLocker snap_locker(m_image_ctx.snap_lock);
+      m_journal->append_op_event(journal_op_tid,journal::EventEntry(journal::ExclusiveLockReleaseEvent()),ctx);
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Add a new ```librbd::Journal::append_lock_release(Context *ctx)``` method and invoke it here.  That method would look similar to the following:

```
template <typename I>
void Journal<I>::append_lock_release_event(Context *ctx) {
  CephContext *cct = m_image_ctx.cct;
  ldout(cct, 10) << this << " " << __func__ << dendl;

  journal::EventEntry event_entry((journal:: ExclusiveLockReleaseEvent()));

  bufferlist bl;
  ::encode(event_entry, bl);

  Future future;
  {
    Mutex::Locker locker(m_lock);
    assert(m_state == STATE_READY);
    future = m_journaler->append(m_tag_tid, bl);
  }
  future.flush(ctx);
}
```

In test/librbd/mock/MockJournal.h, you'll need to add a new mock to match the new method signature:

```
MOCK_METHOD1(append_lock_release_event, void(Context *));
```

In test/librbd/exclusive_lock/test_mock_ExclusiveLock.cc, you'll need to set up the expectation that the ```append_lock_release_event``` method will be called:

```
  void expect_append_lock_release_event(MockImageCtx &mock_image_ctx,
                                        MockJournal &mock_journal, int r) {
    EXPECT_CALL(mock_journal, append_lock_release_event(_))
                  .WillOnce(CompleteContext(r, mock_image_ctx.image_ctx->op_work_queue));
  }
```

In each test, invoke the new ```expect_append_lock_release_event``` method after ```expect_flush_notifies``` (since that is the order in which the methods are invoked).

**Comment by sahithi-rv** [comment]
I think the error is with invoking append_lock_release_event.

**Comment by sahithi-rv** [comment]
Now, I don't get any errors on my local machine when I run RBD_Features=109 ..

**Comment by sahithi-rv** [comment]
In src/test/run-rbd-unit-tests.sh, L#9 and L12: shouldn't "unittest_librbd"  actually be "../unittest_librbd" ?
Otherwise running run-rbd-unit-tests.sh, throws error "file not found". 
Nevertheless on changing this, I get segmentation fault.

**Comment by dillaman** [comment]
@sahithi-rv Negative -- you run it from the "src" directory a la "./test/run-rbd-unit-tests.sh"

File: src/librbd/Journal.h [hunk 13]
```diff
@@ -239,6 +240,7 @@ class Journal {
   };
 
   Journaler *m_journaler;
+  uint64_t m_tag_id;
```
Comment (dillaman) [inline src/librbd/Journal.h]: This doesn't appear to be used -- perhaps pulled in from a bad rebase.

File: src/librbd/exclusive_lock/ReleaseRequest.cc [hunk 14]
```diff
@@ -130,22 +131,51 @@ Context *ReleaseRequest<I>::handle_flush_notifies(int *ret_val) {
...
+       return;
+    }
+
+    using klass = ReleaseRequest<I>;
+    Context *ctx = create_context_callback<
+      klass, &klass::handle_append_journal_event>(this);
+    //append journal event
+    {
+      RWLock::WLocker snap_locker(m_image_ctx.snap_lock);
+      m_image_ctx.journal->append_lock_release_event(ctx);
```
Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Immediate crash dereferencing a null pointer during `unittest_librbd`.  You need to use `m_journal` instead of `m_image_ctx.journal`.

File: src/xxHash [hunk 15]
```diff
@@ -1 +1 @@
-Subproject commit 1f40c6511fa8dd9d2e337ca8c9bc18b3e87663c9
+Subproject commit 44a6297b298e59ab7452defe859f21ed8371aa1c
```
Comment (dillaman) [inline src/xxHash]: This shouldn't be included in the PR

File: src/librbd/exclusive_lock/ReleaseRequest.cc (hunk 14 above)
Comment (sahithi-rv) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: Using m_journal instead of m_image_ctx.journal, I get error at L#87 in test/librbd/exclusive_lock/test_mock_ReleaseRequest.cc. It says expected to be called once but never called.
Am i missing anything? Do I need to invoke anywhere else?

Comment (dillaman) [inline src/librbd/exclusive_lock/ReleaseRequest.cc]: In the SuccessJournalDisabled test, I wouldn't expect ```append_lock_release_event``` to be called (since the journal is disabled).

**Comment by sahithi-rv** [comment]
I do not get error when I run RBD_FEATURES=61 ./unittest_librbd , but i get segfault when I run run-rbd-unit-tests.sh . But I think both are doing pretty much the same thing?

**Comment by dillaman** [comment]
`RBD_FEATURES=61` doesn't test any journaling-related functionality.  `run-rbd-unit-tests.sh` runs `unittest_librbd` with several combinations for `RBD_FEATURES` -- including `RBD_FEATURES=109` which enables support for journaling.

**Comment by sahithi-rv** [comment]
RBD_FEATURES=109 also doesn't throw any error.

**Comment by dillaman** [comment]
Which RBD_FEATURES=x causes the crash?

**Comment by sahithi-rv** [comment]
x = 61 and x=109

**Comment by dillaman** [comment]
Now I am confused since you said 109 doesn't have any issue.

**Comment by sahithi-rv** [comment]
Ok. sorry I have used m_image_ctx first and tested. then I get not errors with running unittest_librbd with RBD_FEATURES= 61, 109. But i got Segfault when I run run-rbd-unit-tests and i found that $i = 61 and $i -109 give error.

**Comment by sahithi-rv** [comment]
When I use m_journal , and run RBD_FEATURES=61 ./unittest_librbd I get this error
err

**Comment by sahithi-rv** [comment]
I want to know, in test_mock_ExclusiveLock.cc L#171 , "EXPECT_CALL(mock_journal, append_lock_release_event(_))" , I am passing just 1 parameter (i.e the second parameter) to append_lock_release_event(). I need to pass 3 arguments right?

**Comment by sahithi-rv** [comment]
Now I tried to free mock_journal in every test in test_mock_ExclusiveLock and test_mock_ReleaseRequest. This does not give me errors as shown above. But it still throws segfault.
Is there a way to debug segfault, here?

**Comment by dillaman** [comment]
`RBD_FEATURES=109 libtool e gdb ./unittest_librbd`

**Comment by sahithi-rv** [comment]
Now, is there no file unittest_librbd? 
It seems to pass all the tests, yet it throws build failure.

**Comment by sahithi-rv** [comment]
When I run  RBD_FEATURES=109 ./unittest_librbd, the tests run successfully till "TestMockImageRefreshRequest.SuccessV2" , but is stuck at "TestMockImageRefreshRequest.SuccessSnapshotV2"

**Comment by sahithi-rv** [comment]
What does it mean append_lock_release_event(_) is not called? I get this error:
err

**Comment by dillaman** [comment]
@sahithi-rv It's hard for me to say since the code available on githut doesn't match up with that error message.  It appears like in your version of the code, for the "UnlockError" test, you had an expectation for "expect_append_lock_release_event", but it was never called before the lock was released.

**Comment by dillaman** [comment]
Closing this PR --  the Jewel version uses journal tag epochs to track the acquire/release of the exclusive lock.
